### ISSUE TITLE
Panic with malformed ntfs in non_resident_value_data_and_position

### ISSUE THREAD

**Issue by Arthur-PI** [ColinFinck/ntfs#20]
Hello, I started fuzzing the crate and found a lot of crashes (panics) and infinite loops. A lot of the panics are caused by unsafe use of `[]` slice operator, sometimes it is out of bounds, some times its a range slice which is not valid like `[start..end]` with `start=320` and `end=307`. I'm gonna try to fix all the panics and write more safe rust code, but I'd like your opinion and if you have a prefered version of how you want to deal with it.

Well, for an example, here is the first bug:
https://github.com/ColinFinck/ntfs/blob/9348d729bcefe0b9473d29b8d7972d5e4038c910/src/attribute.rs#L281-L289
In this lines with the given binary NTFS [test02.zip](https://github.com/ColinFinck/ntfs/files/10338111/test02.zip) it panics because `start` is equal to `320` and `end` to `307` so the range is not valid. I don't understand the deeper problem here but can give a quick fix like this:
```rust
pub(crate) fn non_resident_value_data_and_position(&self) -> Result<(&'f [u8], NtfsPosition)> {
    debug_assert!(!self.is_resident());
    let start = self.offset + self.non_resident_value_data_runs_offset() as usize;
    let end = self.offset + self.attribute_length() as usize;
    let position = self.file.position() + start;
    if let Some(data) = &self.file.record_data().get(start..end) {
        Ok((data, position))
    } else {
        Err(NtfsError::MissingIndexAllocation {position})
    }
}
```

I also don't know which `NtfsError` to use, I put `MissingIndexAllocation` here because that was an easy one and fairly accurate.
I have the PR ready with the the changes induces by the `Result` in the return of the function.

Thanks for your time

**Comment by ColinFinck** [comment]
Hey @Arthur-PI! Thank you for your fuzzing efforts and for opening this bug report. This is a very good catch!

When writing the ntfs crate, my style was to use the `[start..end]` range slice only when `start` and `end` have already undergone extensive validation and are safe to use. An example for this style is https://github.com/ColinFinck/ntfs/blob/9348d729bcefe0b9473d29b8d7972d5e4038c910/src/attribute.rs#L249 in the same file.
This is safe, because `start` and `end` have already been checked in `validate_name_sizes` before: https://github.com/ColinFinck/ntfs/blob/9348d729bcefe0b9473d29b8d7972d5e4038c910/src/attribute.rs#L387-L407
If validation fails, this function also returns a precise error that pinpoints to the exact error type and location in the filesystem.

Offset and length are not necessarily checked and reported individually. Looking at the code again, I also found places where `start` and `end` are put into a `slice.get(start..end)` call, and a single error is returned if `get` doesn't yield a result: https://github.com/ColinFinck/ntfs/blob/9348d729bcefe0b9473d29b8d7972d5e4038c910/src/index_entry.rs#L143-L148
So as you see, there are currently multiple ways to deal with these cases, with room for improvement/unification.

As you ask for my preferred version to deal with such cases:
* Check if there is a suitable error variant. Don't hesitate to add one. Exact error messages are way better than a single message reused for multiple cases.
  In this case, `NtfsError::MissingIndexAllocation` is a specific error when an $INDEX_ALLOCATION attribute was expected but not found. It's unrelated to the reported problem.

* Changing `[start..end]` to `.get(start..end)` is also how I would fix this particular problem.
  One could also check `start` and `end` individually, like it's done in `validate_resident_value_sizes`, but I'd say it's too easy to miss a check here.

* Add an error variant `InvalidNonResidentValueDataRange`, with fields `position`, `range`, and `size`, and return that as the specific error for this case.
  Use `self.position()` for the `position` field, `start..end` for the `range` field, and `self.file.ntfs().file_record_size()` for the size field. This gives all important context to understand and pinpoint the error.
  Doing it like this is also 100% consistent with what's done for `slice.get(start..end)` in index_entry.rs.

* Use `.get(start..end).ok_or(.....)?` instead of an if-else block.
  This doesn't add an indendation level for the happy case, and therefore scales better if we later need to introduce another check with early return in case of an error.

Again, thanks a lot for your help with this!

**Comment by Arthur-PI** [comment]
Thanks you for this very complete response, those explanations are really helpful 👍🏼. 
So I've taking into account what you've said and I made the fix accordingly, here is the new version.
```rust
pub(crate) fn non_resident_value_data_and_position(&self) -> Result<(&'f [u8], NtfsPosition)> {
    debug_assert!(!self.is_resident());
    let start = self.offset + self.non_resident_value_data_runs_offset() as usize;
    let end = self.offset + self.attribute_length() as usize;
    let position = self.file.position() + start;
    let data = &self.file.record_data().get(start..end).ok_or(
        NtfsError::InvalidNonResidentValueDataRange {
            position,
            range: start..end,
            size: self.file.record_data().len(),
        },
    )?;
    Ok((data, position))
}
```
With this new error:
```rust
/// The range of non resident value data when slicing is invalid
InvalidNonResidentValueDataRange {
    position: NtfsPosition,
    range: Range<usize>,
    size: usize,
}
```

Now I have fix for 8 bugs similar to this one, do you prefer one PR for each fix or one big PR with on fix per commit. As you wish. I'll also provide you all the test files that create those crashes (I'll assign a test case by commit) if you want do dig a little bit in why this happens and if the problem is deeper in the parsing process.

Thank you for your work and help.

**Comment by ColinFinck** [comment]
@Arthur-PI Code looks good. Note that the `///` comment for an error variant is parsed by `#[derive(Display)` from `displaydoc::Display`, which is why it contains formatting strings for all fields. Please take care of them as well, so that the context information in the error variant is actually printed.

As we're working asynchronously and I'll be the only one reviewing your PR(s), I think that a single PR with one fix per commit is the best here. Thanks for referencing the test file in each commit - that makes it very easy for me to follow what you've found :)

**Comment by ColinFinck** [comment]
Fixed in #23
